File: 39.组合总和.cpp
/*
 * @lc app=leetcode.cn id=39 lang=cpp
 *
 * [39] 组合总和
 */

// @lc code=start
class Solution {
    vector<vector<int>> ans;
public:
    void dfs(vector<int>& candidates, int startPos, int target, vector<int> &cur) {
        if (target == 0) {
            ans.push_back(cur);
            return;
        }
        for (int i = startPos; i < candidates.size(); i++) {
            if (target - candidates[i] >= 0) {
                cur.push_back(candidates[i]);
                dfs(candidates, i, target - candidates[i], cur);
                cur.pop_back();
            }
        }
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<int> res;
        dfs(candidates, 0, target, res);
        return ans;
    }
};
// @lc code=end

File: 39.组合总和.swift
/*
 * @lc app=leetcode.cn id=39 lang=swift
 *
 * [39] 组合总和
 */

// @lc code=start
class Solution {
    var res = [[Int]]()

    func dfs(_ candidates: [Int], _ startPos: Int, _ target: Int, _ cur: inout [Int]) {
        if target == 0 {
            res.append(cur)
            return
        }
        for idx in startPos..<candidates.count {
            guard target >= candidates[idx] else {
                continue
            }
            cur.append(candidates[idx])
            dfs(candidates, idx, target - candidates[idx], &cur)
            cur.removeLast()
        }
    }

    func combinationSum(_ candidates: [Int], _ target: Int) -> [[Int]] {
        var candidatesCopy = candidates
        candidatesCopy.sort(by: <)
        var cur = [Int]()
        dfs(candidatesCopy, 0, target, &cur)
        return res
    }
}
// @lc code=end